crates/nu-command/src/filters/each.rs

```diff
@@ -238,6 +148,130 @@ with 'transpose' first."#
+            .category(Category::Filters)
+    }
+
+    fn examples(&self) -> Vec<Example> {
+        let stream_test_1 = vec![Value::test_int(2), Value::test_int(4), Value::test_int(6)];
+
+        let stream_test_2 = vec![
+            Value::nothing(Span::test_data()),
+            Value::test_string("found 2!"),
+            Value::nothing(Span::test_data()),
+        ];
+
+        vec![
+            Example {
+                example: "[1 2 3] | each {|e| 2 * $e }",
+                description: "Multiplies elements in the list",
+                result: Some(Value::list(stream_test_1, Span::test_data())),
+            },
+            Example {
+                example: "{major:2, minor:1, patch:4} | values | each {|| into string }",
+                description: "Produce a list of values in the record, converted to string",
+                result: Some(Value::list(
+                    vec![
+                        Value::test_string("2"),
+                        Value::test_string("1"),
+                        Value::test_string("4"),
+                    ],
+                    Span::test_data(),
+                )),
+            },
+            Example {
+                example: r#"[1 2 3 2] | each {|e| if $e == 2 { "two" } }"#,
+                description: "Produce a list that has \"two\" for each 2 in the input",
+                result: Some(Value::list(
+                    vec![Value::test_string("two"), Value::test_string("two")],
+                    Span::test_data(),
+                )),
+            },
+            Example {
+                example: r#"[1 2 3] | enumerate | each {|e| if $e.item == 2 { $"found 2 at ($e.index)!"} }"#,
+                description:
+                    "Iterate over each element, producing a list showing indexes of any 2s",
+                result: Some(Value::list(
+                    vec![Value::test_string("found 2 at 1!")],
+                    Span::test_data(),
+                )),
+            },
+            Example {
+                example: r#"[1 2 3] | each --keep-empty {|e| if $e == 2 { "found 2!"} }"#,
+                description: "Iterate over each element, keeping null results",
+                result: Some(Value::list(stream_test_2, Span::test_data())),
+            },
+        ]
+    }
+
+    fn run_debug(
+        &self,
+        engine_state: &EngineState,
+        stack: &mut Stack,
+        call: &Call,
+        input: PipelineData,
+        debugger: Arc<Mutex<dyn Debugger>>,
+    ) -> Result<PipelineData, ShellError> {
+        run_each!(
+            engine_state,
+            stack,
+            call,
+            input,
+            WithDebug,
+            &Some(debugger.clone())
+        )
+    }
+
+    fn run(
+        &self,
+        engine_state: &EngineState,
+        stack: &mut Stack,
+        call: &Call,
+        input: PipelineData,
+    ) -> Result<PipelineData, ShellError> {
+        run_each!(engine_state, stack, call, input, WithoutDebug, &None)
```

Comment:
why is this a macro and not a function? You could force the function to be inline if you're worried about perf.

Reply:
I tried using a function, but one of the function parameters would have to be `Option<Arc<Mutex<dyn Debugger>>>` or similar. We'd have to then clone it in each loop iteration before passing it to the `eval_block_with_early_return()` even in the non-debug path because of the move closure.